### PR TITLE
drivers: serial: ns16550: Enable simultaneous support of legacy, MMIO and PCIe UART

### PR DESCRIPTION
**Pull request by akanisetti**
Enabled simultaneous support by including a DTS variable named “io-mapped”.
There are 3 possibilities through instance in dtsi file.
Under PCIe, PCIe ns16550.
Under soc and has a variable io-mapped, legacy (IO mapped).
Under soc and don’t have a variable io-mapped, MMIO mapped.
For PCIe instances UART initialization should be done post-kernel as it
depends on PCIe initialization.

Co-authored-by: Najumon BA <najumon.ba@intel.com>
Signed-off-by: Anisetti Avinash Krishna <anisetti.avinash.krishna@intel.com>

### CODE REVIEW

File: drivers/serial/Kconfig.ns16550 [hunk 1]
```diff
@@ -66,6 +84,14 @@ config UART_NS16550_ACCESS_IOPORT
 	  When enabled, NS16550 will not be a memory mapped device. This option
 	  must be selected at SoC/board level if needed.
 
+config UART_NS16550_SIMULT_ACCESS
+	bool
+	default n
```
Comment (jhedberg) [inline drivers/serial/Kconfig.ns16550:89]: You can remove this line (`n` is the "default default")

File: drivers/serial/Kconfig.ns16550 [hunk 2]
```diff
@@ -53,6 +53,24 @@ config UART_NS16550_VARIANT_NS16950
...
+
+config UART_NS16550_PRE_KERNEL
+	bool "ns16550 UART is defined PRE-KERNEL"
+	help
+	  This will set uart driver as pre-kernel
+
+config UART_NS16550_POST_KERNEL
+	bool "ns16550 UART is defined POST-KERNEL"
+	help
+	  This will set uart driver as post-kernel
```
Comment (jhedberg) [inline drivers/serial/Kconfig.ns16550:70]: UART. Also, I'd either use "post-kernel" (human language), or "POST_KERNEL" (the preprocessor macro), but not "POST-KERNEL" (I don't think that exists?)

File: drivers/serial/Kconfig.ns16550 [hunk 3]
```diff
@@ -53,6 +53,24 @@ config UART_NS16550_VARIANT_NS16950
...
+	help
+	  Select boot priority for UART
+
+config UART_NS16550_PRE_KERNEL
+	bool "ns16550 UART is defined PRE-KERNEL"
+	help
+	  This will set uart driver as pre-kernel
+
+config UART_NS16550_POST_KERNEL
+	bool "ns16550 UART is defined POST-KERNEL"
```
Comment (jhedberg) [inline drivers/serial/Kconfig.ns16550:68]: NS16550

File: drivers/serial/Kconfig.ns16550 [hunk 4]
```diff
@@ -66,6 +84,14 @@ config UART_NS16550_ACCESS_IOPORT
 	  When enabled, NS16550 will not be a memory mapped device. This option
 	  must be selected at SoC/board level if needed.
 
+config UART_NS16550_SIMULT_ACCESS
+	bool
+	default n
+	help
+	  When enabled, NS16550 supports IO, MMIO, PCIe uart devices simultaneously.
```
Comment (jhedberg) [inline drivers/serial/Kconfig.ns16550:91]: Here you're spelling it uppercase NS16550, which seems correct. But: UART (not "uart").

File: drivers/serial/Kconfig.ns16550 [hunk 5]
```diff
@@ -66,6 +84,14 @@ config UART_NS16550_ACCESS_IOPORT
...
+	  For IO Mapped instance io-mapped variable need to be added in dtsi
```
Comment (jhedberg) [inline drivers/serial/Kconfig.ns16550:92]: `io-mapped` is a DTS property, not a variable

**Review by jhedberg — CHANGES_REQUESTED** [review]
Some documentation related comments. Technically the PR seems alright to me.

**Review by dcpleung — APPROVED** [review]

File: drivers/serial/uart_ns16550.c [hunk 6]
```diff
@@ -246,13 +226,14 @@ struct uart_ns16550_device_config {
 #if defined(CONFIG_PINCTRL)
 	const struct pinctrl_dev_config *pincfg;
 #endif
+#if defined(CONFIG_UART_NS16550_ACCESS_IOPORT) || defined(CONFIG_UART_NS16550_SIMULT_ACCESS)
```
Comment (jhedberg) [inline drivers/serial/uart_ns16550.c:229]: If `CONFIG_UART_NS16550_ACCESS_IOPORT` is set, then doesn't that mean that every instance is assumed to be using IO port access? In that case this variable is then also unnecessary, i.e. it should only depend on `SIMULT_ACCESS`, right?

**Review by jhedberg — CHANGES_REQUESTED** [review]

**Comment by jhedberg** [comment]
It seems one issue that's popping up, is that the Port IO helpers (`sys_in/out*()` functions) are only defined for ARC and x86 architectures. This leads to compilation errors/warnings now in the driver, unless those calls are #ifdefed away, however there is no generic (non-platform specific) Kconfig option to depend on to do that.

I'd propose to add a new `CONFIG_ARCH_HAS_PORT_IO` Kconfig option which could then be used to set the dependencies correctly. E.g. the `CONFIG_UART_NS16550_ACCESS_IOPORT` and `CONFIG_UART_NS16550_SIMULT_ACCESS` options would both have `depends on ARCH_HAS_PORT_IO`, which would make any compilation issues go away.

@dcpleung @galak @nashif any thoughts on the above?

**Comment by jhedberg** [comment]
> I'd propose to add a new `CONFIG_ARCH_HAS_PORT_IO` Kconfig option which could then be used to set the dependencies correctly. E.g. the `CONFIG_UART_NS16550_ACCESS_IOPORT` and `CONFIG_UART_NS16550_SIMULT_ACCESS` options would both have `depends on ARCH_HAS_PORT_IO`, which would make any compilation issues go away.

Just realised that this will not be enough, since not even the header file prototypes of `sys_in/out()` APIs are available on architectures that do not support Port IO. I wonder if it'd make sense to add static inline functions of those into `arch/common/sys_io.h` with asserts "Not supported on this architecture"?

**Review by jhedberg — APPROVED** [review]
I'll approve this, since I think the readability aspects and more generic handling of Port IO support can be fixed in a later PR.

**Pull request merged**